File: app/src/main/kotlin/io/github/getsixtyfour/openpyn/map/util/CameraUpdateAnimator.java
/*
 * Copyright 2016 ANTONIO CARLON
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.getsixtyfour.openpyn.map.util;

import android.annotation.SuppressLint;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMap.OnCameraIdleListener;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.Collection;

public final class CameraUpdateAnimator implements OnCameraIdleListener {

    private final ArrayList<Animation> mCameraUpdates;

    private final CancelableCallback mCancelableCallback = new CancelableCallback();

    private final Handler mHandler = new Handler();

    private GoogleMap mGoogleMap;

    private OnCameraIdleListener mOnCameraIdleListener;

    private boolean mAnimating;

    @Nullable
    private AnimatorListener mAnimatorListener;

    private boolean mIsRotateGestureEnabled;

    private boolean mIsScrollGestureEnabled;

    private boolean mIsTiltGestureEnabled;

    private boolean mIsZoomControlsEnabled;

    private boolean mIsZoomGestureEnabled;

    public CameraUpdateAnimator(@NonNull GoogleMap googleMap, @NonNull OnCameraIdleListener onCameraIdleListener) {
        this(googleMap, onCameraIdleListener, new ArrayList<>());
    }

    @SuppressLint("LambdaLast")
    public CameraUpdateAnimator(@NonNull GoogleMap googleMap, @NonNull OnCameraIdleListener onCameraIdleListener, @NonNull Collection<Animation> animations) {
        mGoogleMap = googleMap;
        mOnCameraIdleListener = onCameraIdleListener;
        mCameraUpdates = new ArrayList<>(animations);
    }

    private static void startAnimation(@NonNull GoogleMap googleMap, @NonNull Animation animation,
                                       @Nullable GoogleMap.CancelableCallback cancelableCallback) {
        if (animation.isAnimate()) {
            if (cancelableCallback != null) {
                googleMap.animateCamera(animation.getCameraUpdate(), cancelableCallback);
            } else {
                googleMap.animateCamera(animation.getCameraUpdate());
            }
        } else {
            googleMap.moveCamera(animation.getCameraUpdate());
        }
    }

    @Override
    public void onCameraIdle() {
        executeNext();
    }

    public boolean add(@NonNull Animation animation) {
        return mCameraUpdates.add(animation);
    }

    public boolean addAll(@NonNull Collection<Animation> animations) {
        return mCameraUpdates.addAll(animations);
    }

    public void clear() {
        mCameraUpdates.clear();
    }

    public void execute() {
        if (!mCameraUpdates.isEmpty()) {
            setUiSettings();
            onAnimationStart();
            executeNext();
        }
    }

    @Nullable
    public AnimatorListener getAnimatorListener() {
        return mAnimatorListener;
    }

    public void setAnimatorListener(@Nullable AnimatorListener animatorListener) {
        mAnimatorListener = animatorListener;
    }

    public boolean isAnimating() {
        return mAnimating;
    }

    @SuppressWarnings({ "FieldMissingNullable", "AssignmentToNull" })
    public void onDestroy() {
        mHandler.removeCallbacksAndMessages(null);
        mCancelableCallback.setAnimatorListener(null);
        mAnimatorListener = null;
        mOnCameraIdleListener = null;
        mGoogleMap = null;
    }

    private void executeNext() {
        if (mCameraUpdates.isEmpty()) {
            onAnimationEnd();
        } else {
            Animation animation = mCameraUpdates.remove(0);
            if (animation.isCallback()) {
                mCancelableCallback.setAnimation(animation);
                mCancelableCallback.setAnimatorListener(mAnimatorListener);
                if (animation.getDelay() == 0L) {
                    startAnimation(mGoogleMap, animation, mCancelableCallback);
                } else {
                    mHandler.postDelayed(() -> startAnimation(mGoogleMap, animation, mCancelableCallback), animation.getDelay());
                }
            } else {
                if (animation.getDelay() == 0L) {
                    startAnimation(mGoogleMap, animation, null);
                } else {
                    mHandler.postDelayed(() -> startAnimation(mGoogleMap, animation, null), animation.getDelay());
                }
            }
        }
    }

    private void onAnimationEnd() {
        mOnCameraIdleListener.onCameraIdle();
        if (mAnimatorListener != null) {
            mAnimatorListener.onAnimationEnd();
        }
        mGoogleMap.setOnCameraIdleListener(mOnCameraIdleListener);
        UiSettings settings = mGoogleMap.getUiSettings();
        settings.setRotateGesturesEnabled(mIsRotateGestureEnabled);
        settings.setScrollGesturesEnabled(mIsScrollGestureEnabled);
        settings.setTiltGesturesEnabled(mIsTiltGestureEnabled);
        settings.setZoomControlsEnabled(mIsZoomControlsEnabled);
        settings.setZoomGesturesEnabled(mIsZoomGestureEnabled);
        mAnimating = false;
    }

    private void onAnimationStart() {
        if (mAnimatorListener != null) {
            mAnimatorListener.onAnimationStart();
        }
        UiSettings settings = mGoogleMap.getUiSettings();
        settings.setRotateGesturesEnabled(false);
        settings.setScrollGesturesEnabled(false);
        settings.setTiltGesturesEnabled(false);
        settings.setZoomControlsEnabled(false);
        settings.setZoomGesturesEnabled(false);
        mGoogleMap.setOnCameraIdleListener(this);
        mAnimating = true;
    }

    private void setUiSettings() {
        UiSettings settings = mGoogleMap.getUiSettings();
        mIsRotateGestureEnabled = settings.isRotateGesturesEnabled();
        mIsScrollGestureEnabled = settings.isScrollGesturesEnabled();
        mIsTiltGestureEnabled = settings.isTiltGesturesEnabled();
        mIsZoomControlsEnabled = settings.isZoomControlsEnabled();
        mIsZoomGestureEnabled = settings.isZoomGesturesEnabled();
    }

    @SuppressWarnings("PublicInnerClass")
    public interface AnimatorListener {

        /**
         * <p>Notifies the start of the animation.</p>
         */
        void onAnimationStart();

        /**
         * <p>Notifies the end of the animation.</p>
         */
        void onAnimationEnd();

        /**
         * <p>Notifies the finishing of the animation.</p>
         *
         * @param animation The animation which was finished.
         */
        void onAnimationFinish(@NonNull Animation animation);

        /**
         * <p>Notifies the cancellation of the animation.</p>
         *
         * @param animation The animation which was canceled.
         */
        void onAnimationCancel(@NonNull Animation animation);
    }

    @SuppressWarnings("PublicInnerClass")
    public static final class Animation {

        private final CameraUpdate mCameraUpdate;

        private boolean mAnimate;

        private boolean mCallback;

        private boolean mClosest;

        private long mDelay;

        @Nullable
        private Object mTag;

        @Nullable
        private LatLng mTarget;

        public Animation(@NonNull CameraUpdate cameraUpdate) {
            mCameraUpdate = cameraUpdate;
        }

        public boolean isCallback() {
            return mCallback;
        }

        public void setCallback(boolean callback) {
            mCallback = callback;
        }

        @NonNull
        public CameraUpdate getCameraUpdate() {
            return mCameraUpdate;
        }

        public long getDelay() {
            return mDelay;
        }

        public void setDelay(long delay) {
            mDelay = delay;
        }

        @Nullable
        public Object getTag() {
            return mTag;
        }

        public void setTag(@Nullable Object tag) {
            mTag = tag;
        }

        @Nullable
        public LatLng getTarget() {
            return mTarget;
        }

        public void setTarget(@Nullable LatLng target) {
            mTarget = target;
        }

        public boolean isAnimate() {
            return mAnimate;
        }

        public void setAnimate(boolean animate) {
            mAnimate = animate;
        }

        public boolean isClosest() {
            return mClosest;
        }

        public void setClosest(boolean closest) {
            mClosest = closest;
        }
    }

    private static class CancelableCallback implements GoogleMap.CancelableCallback {

        private Animation mAnimation;

        @Nullable
        private AnimatorListener mAnimatorListener;

        CancelableCallback() {
        }

        @Override
        public void onCancel() {
            if (mAnimatorListener != null) {
                mAnimatorListener.onAnimationCancel(mAnimation);
            }
        }

        @Override
        public void onFinish() {
            if (mAnimatorListener != null) {
                mAnimatorListener.onAnimationFinish(mAnimation);
            }
        }

        public void setAnimation(Animation animation) {
            mAnimation = animation;
        }

        public void setAnimatorListener(@Nullable AnimatorListener animatorListener) {
            mAnimatorListener = animatorListener;
        }
    }
}

File: app/src/main/kotlin/io/github/getsixtyfour/openpyn/map/util/CameraUpdateAnimator.kt
package io.github.getsixtyfour.openpyn.map.util

import android.annotation.SuppressLint
import android.os.Handler
import com.google.android.gms.maps.CameraUpdate
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.OnCameraIdleListener
import com.google.android.gms.maps.model.LatLng
import java.util.ArrayList

class CameraUpdateAnimator @SuppressLint("LambdaLast") constructor(
    googleMap: GoogleMap, onCameraIdleListener: OnCameraIdleListener, animations: Collection<Animation>
) : OnCameraIdleListener {

    private val mCameraUpdates: ArrayList<Animation>
    private val mCancelableCallback = CancelableCallback()
    private val mHandler = Handler()
    private var mGoogleMap: GoogleMap?
    private var mOnCameraIdleListener: OnCameraIdleListener?
    var isAnimating: Boolean = false
        private set
    var animatorListener: AnimatorListener? = null
    private var mIsRotateGestureEnabled = false
    private var mIsScrollGestureEnabled = false
    private var mIsTiltGestureEnabled = false
    private var mIsZoomControlsEnabled = false
    private var mIsZoomGestureEnabled = false

    constructor(googleMap: GoogleMap, onCameraIdleListener: OnCameraIdleListener) : this(
        googleMap, onCameraIdleListener, ArrayList<Animation>()
    )

    override fun onCameraIdle() {
        executeNext()
    }

    fun add(animation: Animation): Boolean {
        return mCameraUpdates.add(animation)
    }

    fun addAll(animations: Collection<Animation>): Boolean {
        return mCameraUpdates.addAll(animations)
    }

    fun clear() {
        mCameraUpdates.clear()
    }

    fun execute() {
        if (mCameraUpdates.isNotEmpty()) {
            setUiSettings()
            onAnimationStart()
            executeNext()
        }
    }

    fun onDestroy() {
        mHandler.removeCallbacksAndMessages(null)
        mCancelableCallback.setAnimatorListener(null)
        animatorListener = null
        mOnCameraIdleListener = null
        mGoogleMap = null
    }

    private fun executeNext() {
        if (mCameraUpdates.isEmpty()) {
            onAnimationEnd()
        } else {
            val animation = mCameraUpdates.removeAt(0)
            if (animation.isCallback) {
                mCancelableCallback.setAnimation(animation)
                mCancelableCallback.setAnimatorListener(animatorListener)
                if (animation.delay == 0L) {
                    startAnimation(mGoogleMap!!, animation, mCancelableCallback)
                } else {
                    mHandler.postDelayed({
                        startAnimation(
                            mGoogleMap!!, animation, mCancelableCallback
                        )
                    }, animation.delay)
                }
            } else {
                if (animation.delay == 0L) {
                    startAnimation(mGoogleMap!!, animation, null)
                } else {
                    mHandler.postDelayed(
                        { startAnimation(mGoogleMap!!, animation, null) }, animation.delay
                    )
                }
            }
        }
    }

    private fun onAnimationEnd() {
        mOnCameraIdleListener!!.onCameraIdle()
        if (animatorListener != null) {
            animatorListener!!.onAnimationEnd()
        }
        mGoogleMap!!.setOnCameraIdleListener(mOnCameraIdleListener)
        val settings = mGoogleMap!!.uiSettings
        settings.isRotateGesturesEnabled = mIsRotateGestureEnabled
        settings.isScrollGesturesEnabled = mIsScrollGestureEnabled
        settings.isTiltGesturesEnabled = mIsTiltGestureEnabled
        settings.isZoomControlsEnabled = mIsZoomControlsEnabled
        settings.isZoomGesturesEnabled = mIsZoomGestureEnabled
        isAnimating = false
    }

    private fun onAnimationStart() {
        if (animatorListener != null) {
            animatorListener!!.onAnimationStart()
        }
        val settings = mGoogleMap!!.uiSettings
        settings.isRotateGesturesEnabled = false
        settings.isScrollGesturesEnabled = false
        settings.isTiltGesturesEnabled = false
        settings.isZoomControlsEnabled = false
        settings.isZoomGesturesEnabled = false
        mGoogleMap!!.setOnCameraIdleListener(this)
        isAnimating = true
    }

    private fun setUiSettings() {
        val settings = mGoogleMap!!.uiSettings
        mIsRotateGestureEnabled = settings.isRotateGesturesEnabled
        mIsScrollGestureEnabled = settings.isScrollGesturesEnabled
        mIsTiltGestureEnabled = settings.isTiltGesturesEnabled
        mIsZoomControlsEnabled = settings.isZoomControlsEnabled
        mIsZoomGestureEnabled = settings.isZoomGesturesEnabled
    }

    interface AnimatorListener {
        /**
         *
         * Notifies the start of the animation.
         */
        fun onAnimationStart()

        /**
         *
         * Notifies the end of the animation.
         */
        fun onAnimationEnd()

        /**
         *
         * Notifies the finishing of the animation.
         *
         * @param animation The animation which was finished.
         */
        fun onAnimationFinish(animation: Animation)

        /**
         *
         * Notifies the cancellation of the animation.
         *
         * @param animation The animation which was canceled.
         */
        fun onAnimationCancel(animation: Animation)
    }

    class Animation(val cameraUpdate: CameraUpdate) {
        var isAnimate: Boolean = false
        var isCallback: Boolean = false
        var isClosest: Boolean = false
        var delay: Long = 0
        var tag: Any? = null
        var target: LatLng? = null
    }

    private class CancelableCallback internal constructor() : GoogleMap.CancelableCallback {
        private var mAnimation: Animation? = null
        private var mAnimatorListener: AnimatorListener? = null
        override fun onCancel() {
            if (mAnimatorListener != null) {
                mAnimatorListener!!.onAnimationCancel(mAnimation!!)
            }
        }

        override fun onFinish() {
            if (mAnimatorListener != null) {
                mAnimatorListener!!.onAnimationFinish(mAnimation!!)
            }
        }

        fun setAnimation(animation: Animation?) {
            mAnimation = animation
        }

        fun setAnimatorListener(animatorListener: AnimatorListener?) {
            mAnimatorListener = animatorListener
        }
    }

    companion object {
        private fun startAnimation(
            googleMap: GoogleMap, animation: Animation, cancelableCallback: GoogleMap.CancelableCallback?
        ) {
            if (animation.isAnimate) {
                if (cancelableCallback != null) {
                    googleMap.animateCamera(animation.cameraUpdate, cancelableCallback)
                } else {
                    googleMap.animateCamera(animation.cameraUpdate)
                }
            } else {
                googleMap.moveCamera(animation.cameraUpdate)
            }
        }
    }

    init {
        mGoogleMap = googleMap
        mOnCameraIdleListener = onCameraIdleListener
        mCameraUpdates = ArrayList(animations)
    }
}
